CrossrefReferenceLinkingPlugin.php

```diff
@@ -0,0 +1,474 @@
+	 * Resets the submission data related to Reference Linking Plugin.
+	 * Used every time the citations for a certain publication are imported.
+	 *
+	 * @param $hookName string 'CitationDAO::afterImportCitations'
+	 * @param $params array [
+	 *  @option integer The publication ID for which the citations are imported
+	 * ]
+	 */
+	public function citationsChanged(string $hookName, array $params): bool
+	{
+		$publicationId = $params[0];
+
+		$publication = Repo::publication()->get($publicationId);
+		$submission = Repo::submission()->get($publication->getData('submissionId'));
+
+		if ($submission->getData($this->getCitationsDiagnosticIdSettingName())) {
+			$submission->setData($this->getCitationsDiagnosticIdSettingName(), null);
+			$submission->setData($this->getAutoCheckSettingName(), null);
+			$submission = Repo::submission()->edit($submission, array(), Application::get()->getRequest());
+		}
+		return false;
+	}
+
+	/**
+	 * Get found Crossref references DOIs for the given publication DOI.
+	 */
+	public function getCrossrefReferencesDOIs(Publication $publication): void
+	{
+		$doi = urlencode($publication->getDoi());
+		if (!empty($doi)) {
+			$citationDao = DAORegistry::getDAO('CitationDAO'); /** @var CitationDAO $citationDao */
+			$citations = $citationDao->getByPublicationId($publication->getId())->toAssociativeArray();
+
+			$citationsToCheck = array();
+			foreach ($citations as $citation) { /** @var Citation $citation */
+				if (!$citation->getData($this->getCitationDoiSettingName())) {
+					$citationsToCheck[$citation->getId()] = $citation;
+				}
+			}
+			if (!empty($citationsToCheck)) {
+				$citationsToCheckKeys = array_keys($citationsToCheck);
+
+				$submission = Repo::submission()->get($publication->getData('submissionId'));
+
+				$matchedReferences = $this->_getResolvedRefs($doi, $submission->getData('contextId'));
+				if ($matchedReferences) {
+					$filteredMatchedReferences = array_filter($matchedReferences, function ($value) use ($citationsToCheckKeys) {
+						return in_array($value['key'], $citationsToCheckKeys);
+					});
+
+					foreach ($filteredMatchedReferences as $matchedReference) {
+						$citation = $citationsToCheck[$matchedReference['key']];
+						$citation->setData($this->getCitationDoiSettingName(), $matchedReference['doi']);
+						$citationDao->updateObject($citation);
+					}
+
+					// remove auto check setting
+					$submission->setData($this->getAutoCheckSettingName(), null);
+
+					$submission = Repo::submission()->edit($submission, array(), Application::get()->getRequest());
+				}
+			}
+		}
+	}
+
+	/**
+	 * Insert reference DOI on the citations and article view page.
+	 *
+	 * @param $hookName string Hook name
+	 * @param $params array [
+	 *  @option Citation
+	 *  @option Smarty
+	 *  @option string Rendered smarty template
+	 * ]
+	 */
+	public function displayReferenceDOI(string $hookName, array $params): bool
+	{
+		$citation =& $params[0]['citation'];
+		$smarty =& $params[1];
```

Comment:
I think these variables don't need to be referenced, also, like I commented above, it's good to type hint these "unknown" variables.